apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt

```diff
@@ -95,14 +99,83 @@ fun ApolloClient.Builder.store(store: ApolloStore, writeToCacheAsynchronously: B
       .writeToCacheAsynchronously(writeToCacheAsynchronously)
 }
 
+enum class WatchErrorHandling {
+  THROW_CACHE_ERRORS,
+  THROW_NETWORK_ERRORS,
+  THROW_CACHE_AND_NETWORK_ERRORS,
+  IGNORE_ERRORS
+}
+
 /***
  * Gets the result from the network, then observes the cache for any changes.
  * Overriding the [FetchPolicy] will change how the result is first queried.
+ * Exception are ignored by default, this can be changed by setting [fetchErrorHandling] for the first fetch and [refetchErrorHandling]
+ * for subsequent fetches.
+ */
+@JvmOverloads
+fun <D : Query.Data> ApolloCall<D>.watch(
+    fetchErrorHandling: WatchErrorHandling = WatchErrorHandling.IGNORE_ERRORS,
+    refetchErrorHandling: WatchErrorHandling = WatchErrorHandling.IGNORE_ERRORS,
+): Flow<ApolloResponse<D>> {
+  var data: D? = null
+  return toFlow()
+      .catch {
+        maybeThrow(it, fetchErrorHandling)
+      }.onEach {
+        data = it.data
+      }.onCompletion {
+        emitAll(
+            watch(data)
+                .catch {
+                  maybeThrow(it, refetchErrorHandling)
+                }
+        )
+      }
+}
+
+/**
+ * Observes the cache for the given data. Unlike [watch], no initial request is executed on the network.
+ * The refetch policy set by [refetchPolicy] will be used.
+ * Cache and network exceptions are propagated.
  */
-fun <D : Query.Data> ApolloCall<D>.watch(): Flow<ApolloResponse<D>> {
-  return copy().addExecutionContext(WatchContext(true)).toFlow()
+fun <D : Query.Data> ApolloCall<D>.watch(data: D?): Flow<ApolloResponse<D>> {
+  return copy().addExecutionContext(WatchContext(data)).toFlow()
```

Comment:
Food for thoughts (no action required): instead of adding an execution context here, we could conditionally add the `WatcherInterceptor` so that it's only present in the chain if `watch()` is called.